Enable password recovery when using ILS Authentication

oharacj opened the pull request:
Add handling to enable password recovery when using ILS Authentication method

oharacj committed:
Initial changes to enable password recovery when using ILS Authentication.

oharacj committed:
Initial changes to enable password recovery when using ILS Authentication.
Updated based on the PHP 8.3 test failures.

oharacj committed:
Initial changes to enable password recovery when using ILS Authentication.
Updated based on the PHP 8.3 test failures. When I ran php cbf it made changes to the if statements that caused this failure.

oharacj committed:
Remove the return clause in setFollowupUrlToReferer method.

demiankatz commented on module/VuFind/src/VuFind/Auth/ILSAuthenticator.php:
```diff
@@ -151,7 +151,7 @@ public function encrypt(?string $text)
      * @param ?string $text    The text to be encrypted or decrypted
      * @param bool    $encrypt True if we wish to encrypt text, False if we wish to
-     * decrypt text.
+     *                         decrypt text.
```
It looks like some of the whitespace in the comments got adjusted unnecessarily. Is it possible your IDE has a setting that's getting a little carried away? ;-)

oharacj replied:
My IDE definitely decided everything needed to be justified. I have corrected this.

demiankatz commented on module/VuFind/src/VuFind/Auth/ILSAuthenticator.php:
```diff
@@ -336,8 +338,19 @@ public function storedCatalogLogin()
                 return $patron;
             }
+        } elseif (!empty($user_name)) {
+            $user = $this->getDbService(UserServiceInterface::class)->getUserByUsername($user_name);
```
Should you do this lookup after checking the cache, so you don't make an unnecessary database call?

demiankatz replied:
Also, should you account for the possibility that no matching user is found?

oharacj replied:
should be done

demiankatz commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -186,6 +186,35 @@ protected function processAuthenticationException(AuthException $e)
+        if (str_starts_with($normReferer, $murNorm)) {
+            return;
+        }
+        parent::setFollowupUrlToReferer($allowCurrentUrl, $extras);
```
What is the reason for extending the method in this controller instead of adjusting the base method?

oharacj replied:
added this code to the base method instead of extending.

demiankatz replied:
Looks like this method still exists here; do you plan to remove it now that the logic has been moved?

demiankatz commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -1729,6 +1758,23 @@ public function recoverAction()
         }
+        //ILS Driver:
+        //if the user hasn't logged in yet, but is found by the ILS, call function
+        //getPatronFromBarcode
```
getPatronFromBarcode isn't a standard ILS driver method. Is this something you are planning to add, or is this local code that doesn't apply upstream? If this is something you plan to introduce officially, I'd discourage use of the word "Barcode" since different systems will use different types of identifiers...

oharacj replied:
renamed barcode to username

demiankatz commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -2079,9 +2125,12 @@ public function newPasswordAction()
         $this->getAuthManager()->updateUserVerifyHash($user);
-        // Login
+        if ($followUp = $this->followup()->retrieve('url')) {
+            $newUrl = strstr($followUp, 'Verify', true) . 'Home';
```
I'm not sure I understand the purpose of this; maybe a comment is in order.

oharacj replied:
added a comment. Happy to answer any other questions about this or taking ANY suggestions. Image is before this code block.

![image](https://github.com/user-attachments/assets/0810c1c5-9879-4652-8ff9-665de2239daf)

demiankatz replied:
Thanks, that makes sense, though I'm not sure I understand exactly what `$followUpUrl = strstr($followUp, 'Verify', true) . 'Home';` is trying to do. I wonder if it might be better to do something like `$followUpUrl = str_contains($followUp, 'Verify') ? $this->url()->fromRoute('home') : $followUp;` (so we're using the router instead of string manipulation to pick the destination URL... but maybe I'm misunderstanding which target route is actually desired).

demiankatz reviewed (changes requested):
Thanks, @oharacj! See below for a few initial thoughts and questions.

EreMaijala commented on module/VuFind/src/VuFind/Auth/ILS.php:
```diff
@@ -113,6 +113,23 @@ public function authenticate($request)
+     *
+     * @return bool
+     */
+    public function supportsPasswordRecovery()
```
This needs a counterpart in MultiILS that supports checking the selected target. For inspiration: 

- https://github.com/NatLibFi/NDL-VuFind2/blob/dev/themes/finna2/templates/Auth/MultiILS/login.phtml#L64 
- https://github.com/NatLibFi/NDL-VuFind2/blob/dev/themes/finna2/js/finna-layout.js#L372
- https://github.com/NatLibFi/NDL-VuFind2/blob/dev/module/Finna/src/Finna/Auth/ILSFinna.php#L53

oharacj replied:
My initial work is only supposed to be for the ILS auth. MultiILS is something I'll look at in the future but I'm not sure it's part of the scope of this work. Is this piece a necessity? If so, I can start working on this and I'll resubmit when I have it done. I only have one ILS and I can't really test but I'm happy to look into it.

demiankatz replied:
You could potentially use the Demo driver as a second ILS for testing purposes, if that helps! Let me know if you need more details/help to get that set up.

EreMaijala commented on module/VuFind/src/VuFind/Auth/ILS.php:
```diff
@@ -113,6 +113,23 @@ public function authenticate($request)
+    {
+        $driver = $this->getCatalog()->getDriver();
+        if (
+            method_exists($driver, 'changePassword')
```
I'd suggest a new method for updating the password during recovery, since changePassword requires the patron from patronLogin method. Our custom code uses this:
https://github.com/NatLibFi/NDL-VuFind2/blob/dev/module/Finna/src/Finna/ILS/Driver/KohaRest.php#L792

oharacj replied:
I've updated the code to use a resetPassword method for this. I've added my resetPassword to the SierraRest Driver.

EreMaijala reviewed (changes requested):
This should work without an existing VuFind account. My proposal would be to add another ILS driver method for looking up the patron with email address and card number. Since we already have a local implementation that works across ILS systems, our experience is that this allows recovery with any ILS that can do the lookup and update the password. See https://github.com/NatLibFi/NDL-VuFind2/blob/dev/module/Finna/src/Finna/ILS/Driver/KohaRest.php#L747 for our implementation with Koha.

The recovery token data and any other information needed (e.g. target ILS for MultiILS) from the lookup method can be stored in access_token table with a random hash as the id. This hash should be included in the recovery link in the email along. 

Since this functionality is somewhat different from internal password recovery, it might make sense to separate the implementation. That's why we chose locally to use different method names etc.

oharacj committed:
Updated based on Demian's comments

Updated comments, line endings, follow-up edits have been put in abstractbase, explained the redirect.

oharacj committed:
Add getPatronFromUsername to SierraRest driver.

oharacj committed:
Add getPatronFromUsername to SierraRest driver.
updated formatting.

oharacj committed:
Add getPatronFromUsername to SierraRest driver.
added trailing comma to return array

oharacj committed:
Add getPatronFromUsername to SierraRest driver.
Object-Oriented Programming is a thing.

oharacj commented:
> This should work without an existing VuFind account. My proposal would be to add another ILS driver method for looking up the patron with email address and card number. Since we already have a local implementation that works across ILS systems, our experience is that this allows recovery with any ILS that can do the lookup and update the password. See https://github.com/NatLibFi/NDL-VuFind2/blob/dev/module/Finna/src/Finna/ILS/Driver/KohaRest.php#L747 for our implementation with Koha.

**I have an ILS method called getPatronFromUsername which provides this function exactly. It has been added to the SierraRest driver.**

> 
> The recovery token data and any other information needed (e.g. target ILS for MultiILS) from the lookup method can be stored in access_token table with a random hash as the id. This hash should be included in the recovery link in the email along.
> 
> Since this functionality is somewhat different from internal password recovery, it might make sense to separate the implementation. That's why we chose locally to use different method names etc.

**I am partially utilising the internal password recovery methods in order to keep from duplicating functionality**

demiankatz commented on config/vufind/SierraRest.ini:
```diff
@@ -59,6 +59,9 @@ use_prefixed_ids = false
 password_field = "pin"
+; Some library systems use a four digit access pin. Others prefer to have alpha numeric
+; passwords. set digits_only to true for pin functionality. Default is false (alpha-num).
+digits_only = true
```
There are [higher-level password rules defined in config.ini](https://github.com/vufind-org/vufind/blob/dev/config/vufind/config.ini#L597-L607). I wonder if it would be beneficial to make this setting more parallel with those (especially if we could refactor code to make the password validation logic reusable -- I haven't inspected to see how feasible this is, but it seems likely to be an option).

Alternatively, if there are really only two different options in Sierra, maybe it would be more clear to call this setting `four_digit_pin` (to account for both the length restriction and the content restriction), or to have a setting like `password_mode = pin|password` (if we want to allow for more possibilities in the future than a binary setting can provide.

demiankatz replied:
Just checking to see if you have any thoughts on this point, since there hasn't been further discussion on it since October. (Also happy to hear @EreMaijala's thoughts, if any).

EreMaijala replied:
I think we should use the same rules as when changing a password, i.e. the [changePassword] section in ILS driver's ini:

```
; Uncomment the following lines to enable password (PIN) change
;[changePassword]
; PIN change parameters. The default limits are taken from the interface documentation.
;minLength = 4
;maxLength = 4
; See the password_pattern/password_hint settings in the [Authentication] section
; of config.ini for notes on these settings. When set here, these will override the
; config.ini defaults when Sierra is used for authentication.
;pattern = "numeric"
;hint = "Your optional custom hint can go here."

```

EreMaijala replied:
(Obviously Voyager shouldn't be mentioned here)

oharacj replied:
I'm still not sure how I've managed to miss this section of the ini so many times. I had to have looked over it hundreds of times without seeing that there was a [changePassword] section. I'll remove the custom stuff I did but I'll also have to make a change to the changePassword function in the SierraRest driver in order to remove the hard-coded digits only section of password change.

demiankatz commented on module/VuFind/src/VuFind/ILS/Driver/SierraRest.php:
```diff
@@ -1640,6 +1682,21 @@ public function getMyFines($patron)
     }
 
+    /**
+     * Recover user's password with a token from getPasswordRecoveryToken
```
Where is getPasswordRecoveryToken defined?

oharacj replied:
I copied the initial comment block from suggestions provided by @EreMaijala but eventually decided to use the existing functionality. I just forgot to update the comment.

demiankatz commented on module/VuFind/src/VuFind/Auth/ILSAuthenticator.php:
```diff
@@ -309,11 +309,12 @@ public function getStoredCatalogCredentials()
      * ones.
      *
-     * Returns associative array of patron data on success, false on failure.
+     * @param $user_name - the username/barcode for ILS password reset
```
Minor points of style: I'd suggest `$username` or `$userName` instead of `$user_name` -- we rarely use snake_case in our code, so the other options would be more consistent. Also, you don't need the "-" separator in this comment.

demiankatz reviewed (changes requested):
Thanks for the progress, @oharacj -- see below for a few more tips, questions and ideas. :-)

oharacj committed:
Multi-ILS Driver work and bug fixes

oharacj committed:
Fixed some encoding and comment issues

oharacj committed:
removed comment block in .js file

oharacj committed:
added displayILSPasswordRecoveryLink to the top exported comment

oharacj committed:
PHP 8.3 updates and space fixes

oharacj committed:
PHP 8.3 updates and space fixes

oharacj committed:
PHP 8.3 double quotes in template

oharacj committed:
PHP 8.3 double quotes in template

demiankatz commented on module/VuFind/src/VuFind/Auth/ILSAuthenticator.php:
```diff
@@ -29,9 +29,8 @@
 
 namespace VuFind\Auth;
 
-use Laminas\Config\Config;
```
I'm a little confused by the Laminas config changes showing up in this PR; that's a change made in the dev branch after this PR was opened. It looks like somehow some parts of those changes made it here, even though the whole upstream dev branch hasn't been merged in.

I think I'm going to pause my review until this is resolved, since it looks like the code is a little bit out of sync. Do you mind merging the upstream dev branch into this PR and pushing up those changes? I'll continue reviewing once everything is up to date again. :-)

oharacj replied:
I believe this should be merged in as of 12/23/2024

demiankatz reviewed (changes requested):
Thanks for the ongoing work, @oharacj -- I started my review this morning but ran into some questions that I think will be resolved by merging the dev branch into this PR, so I decided to stop until that's done to be sure I'm reviewing everything in an up-to-date and working state. Please let me know if you run into any trouble with this!

oharacj committed:
Merge branch 'dev' into ilsPasswordRecovery

oharacj committed:
Remove hard-coded pin length in changePassword function

oharacj committed:
Update settings file to remove redundant setting.

demiankatz commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -2238,9 +2265,9 @@ protected function setUpAuthenticationFromRequest()
         if (!empty($method)) {
-            $this->getAuthManager()->setAuthMethod($method);
-        }
+        }    $this->getAuthManager()->setAuthMethod($method);
```
There seems to be a formatting problem here -- did this brace get moved up a line somehow?

demiankatz commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -2066,18 +2088,23 @@ public function newPasswordAction()
         $this->getAuthManager()->updateUserVerifyHash($user);
-        // Login
+        if ($followUp = $this->followup()->retrieve('url')) {
+            //This exists because the followupURL gets set to Verify which returns
+            //an error message due to trying to check the hash a second time
```
Minor style suggestion -- add spaces after `//`:

```suggestion
            // This exists because the followupURL gets set to Verify which returns
            // an error message due to trying to check the hash a second time
```

demiankatz commented on themes/bootstrap5/templates/myresearch/newpassword.phtml:
```diff
@@ -23,6 +23,9 @@
     <input type="hidden" value="<?=$this->escapeHtmlAttr($this->hash) ?>" name="hash">
     <input type="hidden" value="<?=$this->escapeHtmlAttr($this->username) ?>" name="username">
     <input type="hidden" value="<?=$this->escapeHtmlAttr($this->auth_method) ?>" name="auth_method">
+        <?php if (!empty($this->target)): ?>
+            <input type="hidden" value="<?=$this->escapeHtmlAttr($this->target) ?>" name="target">
+        <?php endif; ?>
```
This seems to have some extra indentation:

```suggestion
    <?php if (!empty($this->target)): ?>
      <input type="hidden" value="<?=$this->escapeHtmlAttr($this->target) ?>" name="target">
    <?php endif; ?>
```

demiankatz reviewed (changes requested):
I didn't have time for a full review yet, but I noticed one thing that looks like a problem and had a couple of very minor style suggestions that you can apply by just committing my suggestions from within the comments below. :-)

oharacj committed:
updating indentation

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

oharacj committed:
update spacing in comment

Co-authored-by: Demian Katz <demian.katz@villanova.edu>

oharacj committed:
update spacing and indents

oharacj committed:
update spacing and indents

demiankatz commented on module/VuFind/src/VuFind/Auth/ILS.php:
```diff
@@ -178,6 +200,55 @@ public function updatePassword($request)
+     * @throws AuthException
+     * @throws \Exception
+     */
+    public function newPassword($request)
```
I'm not entirely sure (without spending more time digging deeper into the code and the history) why we need to split updatePassword into two methods... but if this is unavoidable, I wonder if `resetPassword` would be a better name than `newPassword` since it's more verb-based. Also, if we make this change here, do we need to make parallel changes to `VuFind\Auth\Database` since it is currently implemented to do password resets using only updatePassword, and some of the other code changes here might conflict with that?

EreMaijala replied:
I'm pretty sure that updatePassword is able to do everything needed, so this method shouldn't be needed.

demiankatz commented on module/VuFind/src/VuFind/Auth/Manager.php:
```diff
@@ -204,6 +204,19 @@ public function supportsRecovery($authMethod = null)
+     *
+     * @return bool
+     */
+    public function multiSupportsPasswordRecovery($target)
```
Might it be possible to add a second optional argument to supportsPasswordRecovery instead of a whole new method here? Or is this really unavoidable?

demiankatz commented on module/VuFind/src/VuFind/Controller/AbstractBase.php:
```diff
@@ -712,6 +712,14 @@ protected function setFollowupUrlToReferer(bool $allowCurrentUrl = true, array $
         if ($mrhuNorm === $refererNorm) {
             return;
         }
+        // if  is the stored current url of the lightbox
```
I'm not sure I understand this comment, and `if  is` suggests to me that maybe there are extra or missing words.

demiankatz commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -1725,6 +1725,23 @@ public function recoverAction()
         } elseif ($username = $this->params()->fromPost('username')) {
             $user = $userService->getUserByUsername($username);
         }
+        //ILS Driver:
+        //if the user hasn't logged in yet, but is found by the ILS, call function
+        //getPatronFromUsername
```
Another minor style tweak that you can commit directly from this comment:

```suggestion
        // ILS Driver:
        // if the user hasn't logged in yet, but is found by the ILS, call function
        // getPatronFromUsername
```

demiankatz commented on themes/bootstrap5/templates/myresearch/newpassword.phtml:
```diff
@@ -23,6 +23,9 @@
     <input type="hidden" value="<?=$this->escapeHtmlAttr($this->username) ?>" name="username">
     <input type="hidden" value="<?=$this->escapeHtmlAttr($this->auth_method) ?>" name="auth_method">
+    <?php if (!empty($this->target)): ?>
+      <input type="hidden" value="<?=$this->escapeHtmlAttr($this->target) ?>" name="target">
```
Would it make sense to use a more descriptive name here, like `target_ils`? At a glance, it's not immediately obvious to me what target means, so a little context might help.

demiankatz commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -1766,14 +1783,18 @@ protected function sendRecoveryEmail(UserEntityInterface $user, $config)
                     $method = $this->getAuthManager()->getAuthMethod();
+                    // If target exists create query string to include it as part of reset url
+                    $target = $this->getRequest()->getQuery('target') ? '&target='
+                        . $this->getRequest()->getQuery('target') : null;
```
Since you concatenate this to a string down below, I don't think you ever want it to be null -- should probably be empty string instead. Also, it's probably best to encode the value for safety. And it might make sense to refactor to avoid calling getQuery so many times, e.g.:

```
$target = $this->getRequest()->getQuery('target');
$targetParam = $target ? '&target=' . urlencode($target) : '';
```

Then down below, you can use $targetParam in your concatenation, and you can assign $target to $view->target instead of fetching the value yet another time.

demiankatz reviewed (changes requested):
@oharacj, I've given this another relatively high-level review (since much of this code is outside of my area of expertise, it would take me longer to really refresh all my memory on the details, so I'm just trying to chip away at issues while waiting for others with more experience to find time to review more deeply). Please feel free to ignore any comments that are not relevant due to my naivete. I hope the bombardment of small reviews is not a nuisance! ;-)

EreMaijala commented on module/VuFind/src/VuFind/Auth/ILSAuthenticator.php:
```diff
@@ -329,8 +330,22 @@ public function storedCatalogLogin()
                 $this->ilsAccount[$username] = $patron;
                 return $patron;
             }
+        } elseif (!empty($userName)) {
```
What's this for? I don't think this should be needed for password recovery.

EreMaijala commented on module/VuFind/src/VuFind/Auth/Manager.php:
```diff
@@ -670,6 +683,22 @@ public function updatePassword($request)
+     * @throws AuthException
+     * @return UserEntityInterface Updated user entity.
+     */
+    public function newPassword($request)
```
This should not be necessary either.

EreMaijala commented on module/VuFind/src/VuFind/Auth/MultiILS.php:
```diff
@@ -123,4 +124,24 @@ public function setCatalog(\VuFind\ILS\Connection $connection)
+        $driver = $catalog->getDriver();
+        $targetDriver = $driver;
+        if ($target != null) {
+            $targetDriver = $driver->getDriverFromTarget($target);
```
It's better to leave the details inside MultiBackend and just pass enough information to it for it to be able to choose the correct driver. Something like:

```
    public function ilsSupportsPasswordRecovery($target)
    {
        $catalog = $this->getCatalog();
        $recoveryConfig = $catalog->checkFunction(
            'recoverPassword',
            ['cat_username' => "$target.123"]
        );
        return $recoveryConfig ? true : false;
    }

```
Admittedly not exactly beautiful, and it's my intention make it possible to just indicate the target more elegantly. Somewhere in future.

EreMaijala commented on module/VuFind/src/VuFind/Controller/MyResearchController.php:
```diff
@@ -1725,6 +1725,23 @@ public function recoverAction()
+        //ILS Driver:
+        //if the user hasn't logged in yet, but is found by the ILS, call function
+        //getPatronFromUsername
+        if (!$user && $this->formWasSubmitted() && !empty($username)) {
```
I think this, as opposed to updatePassword, should be split to ILS-specific part. And because the ILS's have different functionality, we can't rely on always being able to fetch a user by username. It could also be email address.
It might also make sense to not even try to retrieve a full user in this situation, but just whatever information is needed when the password is reset. In some cases it could be that the ILS supports password reset by way of a recovery token instead of username, but e.g. for Koha we should just need patron id (borrowernumber).

EreMaijala commented on themes/bootstrap3/templates/myresearch/newpassword.phtml:
```diff
@@ -18,11 +18,14 @@
   <div class="error"><?=$this->transEsc('recovery_user_not_found') ?></div>
 <?php else: ?>
-  <form id="newpassword" class="form-new-password" action="<?=$this->url('myresearch-newpassword') ?>" method="post" data-toggle="validator">
+  <form id="newpassword" class="form-new-password" action="<?=$this->url('myresearch-newpassword') ?>" method="post" data-bs-toggle="validator">
```
There should be data-toggle, no data-bs-toggle in bootstrap3

EreMaijala commented:
Superseded by #4332.
